include/caffeine/Memory/MemHeap.h

```diff
@@ -27,6 +27,16 @@ class Solver;
  */
 enum class AllocationKind { Alloca, Malloc, Global };
 
+/**
+ * Permisisons for an allocation
+ *
+ * This is used to indicate what kind of operations a user is allowed
+ * to perfom on the requested allocation. As an example, for an allocation
+ * with ReadWrite permissions a user could can read and modify the underlying
+ * memory. For a Read, a user can read the data, but cannot modify the buffer.
+ */
+enum class AllocationPermissions { None, Read, Write, ReadWrite };
```

Comment:
If we make this a bitfield then you can easily do `is_readable` or `is_writable checks.

```suggestion
enum class AllocationPermissions { 
  None = 0,
  Read = 1,
  Write = 2, 
  ReadWrite = 3
};
```

I do realize this doesn't change the semantics at all but it makes the values explicit so that they can be relied upon :)